### PR TITLE
AffinePoint in LookupTable + NafLookupTable for Shamir's trick

### LINKED ISSUE #13: Add w-NAF scalar multiplication
In the continuation of #1, it may be interesting to include support for w-NAF scalar decomposition for faster scalar multiplication of group elements.

### LINKED ISSUE #14: Lookup tables with AffinePoint for Projective scalar multiplication
Mixed addition (`AffinePoint` + `ProjectivePoint`) is about 88% faster than regular `ProjectivePoint` addition.
This could lead to beneficial gains in scalar multiplication, by storing in the `LookupTable` multiples of the base point in Affine coordinates, rather than Projective coordinates. This can be done efficiently by creating first the array of `ProjectivePoint` multiples, and then doing a batch normalization to convert them all in Affine coordinates with 1 field inversion.

Unfortunately, we currently have a relatively expensive `Fp6` inversion which makes this approach only beneficial for scalar multiplication with the hardcoded basepoint. But it may be worth investigating, and keep it under the hood in case inversion is improved.

### PR DESCRIPTION
**Pull request by Nashtare**
This PR the internal type of points in `LookupTable` from `ProjectivePoint` to `AffinePoint` to benefit from faster mixed-addition. The performance difference is more important for multiplication with the hardcoded `BASEPOINT_TABLE`, about 13% faster.

It also integrates w-NAF conversion of Scalar elements for efficient Shamir's trick, useful in ECDSA or Schnorr signature verification.

The change to `AffinePoint` as internal type of the LookupTable requires the use of `lazy_static!` to construct the `BASEPOINT_TABLE`.

closes #13 
closes #14

### CODE REVIEW

**Comment by Nashtare** [comment]
NAF based scalar mult with a window size of 5 seems to be even more efficient than the regular Lookup based scalar mult, hence relying on this for "vartime" methods.

File: src/curve.rs [hunk 1]
```diff
@@ -965,62 +976,126 @@ impl ProjectivePoint {
...
+        let by_lhs_digits = Scalar::bytes_to_wnaf_vartime(by_lhs, 5);
+        let by_rhs_digits = Scalar::bytes_to_wnaf_vartime(by_rhs, 5);
+
+        // We skip unset digits
+        let mut i: usize = 255;
+        for j in (0..256).rev() {
+            i = j;
+            if by_lhs_digits[i] != 0 || by_rhs_digits[i] != 0 {
+                break;
+            }
```
Comment (baumbata) [inline src/curve.rs:994]: Would it be slightly better if `i = j` is only done inside the `if`?

File: src/curve.rs [hunk 2]
```diff
@@ -902,58 +911,60 @@ impl ProjectivePoint {
...
+                Ordering::Greater => acc += &table.get_point(digits[i] as usize),
+                Ordering::Less => acc -= &table.get_point(-digits[i] as usize),
+                Ordering::Equal => (),
+            };
+
+            if i == 0 {
+                break;
             }
+
+            i -= 1;
```
Comment (baumbata) [inline src/curve.rs:940]: Is it more efficient to do it this way than to have e.g. `for i = (0..i+1).rev()`?

**Review by baumbata — APPROVED** [review]
Looks good to me.

File: src/curve.rs (hunk 2 above)
Comment (Nashtare) [reply src/curve.rs:940]: I believe it's the same, I'm afraid of readability here as `i` refers to a loop indexed by `i`, so I instead used `for j in (0..i + 1).rev()`.

**Pull request merged**